[Inliner] Fix bug when propagating poison generating return attributes

goldsteinn opened the pull request:
[Inliner] Fix bug when propagating poison generating return attributes
Poison generating return attributes can't be propagated the same as
others, as they can change the behavior of other uses and/or create UB
where it otherwise wouldn't have occurred.

For example:
```
define nonnull ptr @foo() {
    %p = call ptr @bar()
    call void @use(ptr %p)
    ret ptr %p
}
```

If we inline `@foo` and propagate `nonnull` to `@bar`, it could change
the behavior of `@use` as instead of taking `null`, `@use` will
now be passed `poison`.

This can be even worth in a case like:
```
define nonnull ptr @foo() {
    %p = call noundef ptr @bar()
    ret ptr %p
}
```

Where propagating `nonnull` to `@bar` will cause UB on `null` return
of `@bar` (`noundef` + `poison`) where it previously wouldn't
have occurred.

To fix this, we only propagate poison generating return attributes if
either 1) The only use of the callsite to propagate too is return and
the callsite to propagate too doesn't have `noundef`. Or 2) the
callsite to be be inlined has `noundef`.

The former case ensures no new UB or `poison` values will be
added. The latter is UB anyways if the value is `poison` so we can go
ahead without worrying about behavior changes.

goldsteinn committed:
[Inliner] Add some additional tests for progagating attributes before inlining; NFC

goldsteinn commented:
Compile time impact: https://llvm-compile-time-tracker.com/compare.php?from=825df6cfeae3b3aeaa3e8a1882c564bc50d99be5&to=4ccda9426371d37d57042058992931e3bd1ea9e3&stat=instructions:u

Some regression for ThinLTO/FullLTO, minimal for normal O3 build.

RKSimon commented:
> Split some tests for c/cpp; NFC

Why was this necessary?

nikic commented:
It occurs to me that the current return attribute propagation is currently buggy for poison-generating attributes: https://llvm.godbolt.org/z/x8n18q9Mj

In this case the argument to use() will now be poison as well, while before inlining only the return value was poison.

This code needs to distinguish poison and UB generating attributes.

goldsteinn commented:
> It occurs to me that the current return attribute propagation is currently buggy for poison-generating attributes: https://llvm.godbolt.org/z/x8n18q9Mj
> 
> In this case the argument to use() will now be poison as well, while before inlining only the return value was poison.
> 
> This code needs to distinguish poison and UB generating attributes.

Good catch. I think this means basically `nonnull`, `noundef`, and `align` can only be propagated if there are no other uses in to-be-inlined function.
That sound right or do you see any more robust way forward?

nikic commented:
> > It occurs to me that the current return attribute propagation is currently buggy for poison-generating attributes: https://llvm.godbolt.org/z/x8n18q9Mj
> > In this case the argument to use() will now be poison as well, while before inlining only the return value was poison.
> > This code needs to distinguish poison and UB generating attributes.
> 
> Good catch. I think this means basically `nonnull`, `noundef`, and `align` can only be propagated if there are no other uses in to-be-inlined function. That sound right or do you see any more robust way forward?

Limiting poison-generating attributes to one-use while keeping the rest of the logic (with guarantee-to-transfer) would be the simplest way to make the existing code correct. There are two additional ways to generalize:

 * If the return is also noundef, we don't need one-use.
 * Just one-use is enough for poison-generating, we don't need guaranteed-to-transfer, BUT: we need to be careful about an implicit "use" in the call itself. That is, if the call we're transfering to is noundef and we only check one-use but not guaranteed-to-transfer, we would convert poison into UB at that point.

goldsteinn commented:
> > > It occurs to me that the current return attribute propagation is currently buggy for poison-generating attributes: https://llvm.godbolt.org/z/x8n18q9Mj
> > > In this case the argument to use() will now be poison as well, while before inlining only the return value was poison.
> > > This code needs to distinguish poison and UB generating attributes.
> > 
> > 
> > Good catch. I think this means basically `nonnull`, `noundef`, and `align` can only be propagated if there are no other uses in to-be-inlined function. That sound right or do you see any more robust way forward?
> 
> Limiting poison-generating attributes to one-use while keeping the rest of the logic (with guarantee-to-transfer) would be the simplest way to make the existing code correct. There are two additional ways to generalize:
> 
> * If the return is also noundef, we don't need one-use.
Is that right for `nonnull`?

I.e:
```
define noundef nonnull ptr @foo() {
   %b = call ptr @bar()
   call void @use(ptr %p) willreturn nounwind
   ret ptr %b
}
```
If we add `nonnull` to `@bar` during inlining it can still make `%p` poison for its use in `@use`. I get it will trigger proper UB at the return b.c the `noundef`, but is it okay to potentially trigger it early at the call to `@use`?


> * Just one-use is enough for poison-generating, we don't need guaranteed-to-transfer, BUT: we need to be careful about an implicit "use" in the call itself. That is, if the call we're transfering to is noundef and we only check one-use but not guaranteed-to-transfer, we would convert poison into UB at that point.

Oh I see, you mean something like:
```
define nonnull ptr @foo() {
   %b = call noundef ptr @bar()
   ret ptr %b
}
```

So it would convert `poison` ret to full UB if we transfer `nonnull`.
The otherway around too I guess i.e if we have:
```
define noundef ptr @foo() {
   %b = call nonull ptr @bar()
   ret ptr %b
}
```
We also need to be careful about transferring the `noundef`.

I think with one-use + guranteed to transfer (the latter is a precondition for transferring at all)
in the latter we would get UB either way, but we are moving to the point up a tiny bit.
I don't understand how this could ever be okay in the former case though? Isn't the former case
just straight up converting `poison` to UB which is a no-go?

nikic commented:
> ```
> define noundef nonnull ptr @foo() {
>    %b = call ptr @bar()
>    call void @use(ptr %p) willreturn nounwind
>    ret ptr %b
> }
> ```
> 
> If we add `nonnull` to `@bar` during inlining it can still make `%p` poison for its use in `@use`. I get it will trigger proper UB at the return b.c the `noundef`, but is it okay to potentially trigger it early at the call to `@use`?

Yes, this is fine, because the UB will be hit anyway. We can move it backwards, as long as it's guaranteed to execute.

> > * Just one-use is enough for poison-generating, we don't need guaranteed-to-transfer, BUT: we need to be careful about an implicit "use" in the call itself. That is, if the call we're transfering to is noundef and we only check one-use but not guaranteed-to-transfer, we would convert poison into UB at that point.
> 
> Oh I see, you mean something like:
> 
> ```
> define nonnull ptr @foo() {
>    %b = call noundef ptr @bar()
>    ret ptr %b
> }
> ```
> 
> So it would convert `poison` ret to full UB if we transfer `nonnull`. The otherway around too I guess i.e if we have:
> 
> ```
> define noundef ptr @foo() {
>    %b = call nonull ptr @bar()
>    ret ptr %b
> }
> ```
> 
> We also need to be careful about transferring the `noundef`.
>
> I think with one-use + guranteed to transfer (the latter is a precondition for transferring at all) in the latter we would get UB either way, but we are moving to the point up a tiny bit. I don't understand how this could ever be okay in the former case though? Isn't the former case just straight up converting `poison` to UB which is a no-go?

You are right, the former case isn't correct even with one-use to guaranteed-to-transfer. That case would only work if we also had noundef on the foo return value.

goldsteinn commented:
> > ```
> > define noundef nonnull ptr @foo() {
> >    %b = call ptr @bar()
> >    call void @use(ptr %p) willreturn nounwind
> >    ret ptr %b
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If we add `nonnull` to `@bar` during inlining it can still make `%p` poison for its use in `@use`. I get it will trigger proper UB at the return b.c the `noundef`, but is it okay to potentially trigger it early at the call to `@use`?
> 
> Yes, this is fine, because the UB will be hit anyway. We can move it backwards, as long as it's guaranteed to execute.
> 
> > > * Just one-use is enough for poison-generating, we don't need guaranteed-to-transfer, BUT: we need to be careful about an implicit "use" in the call itself. That is, if the call we're transfering to is noundef and we only check one-use but not guaranteed-to-transfer, we would convert poison into UB at that point.
> > 
> > 
> > Oh I see, you mean something like:
> > ```
> > define nonnull ptr @foo() {
> >    %b = call noundef ptr @bar()
> >    ret ptr %b
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So it would convert `poison` ret to full UB if we transfer `nonnull`. The otherway around too I guess i.e if we have:
> > ```
> > define noundef ptr @foo() {
> >    %b = call nonull ptr @bar()
> >    ret ptr %b
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > We also need to be careful about transferring the `noundef`.
> > I think with one-use + guranteed to transfer (the latter is a precondition for transferring at all) in the latter we would get UB either way, but we are moving to the point up a tiny bit. I don't understand how this could ever be okay in the former case though? Isn't the former case just straight up converting `poison` to UB which is a no-go?
> 
> You are right, the former case isn't correct even with one-use to guaranteed-to-transfer. That case would only work if we also had noundef on the foo return value.

@nikic, on the former case, could we just strip `noundef` from the ret attributes?
Just a hunch, but think `nonnull`, `align`,... are generally more useful.

goldsteinn committed:
[Inliner] Fix bug when propagating poison generating return attributes

Poison generating return attributes can't be propagated the same as
others, as they can change the behavior of other uses and/or create UB
where it otherwise wouldn't have occurred.

For example:
```
define nonnull ptr @foo() {
    %p = call ptr @bar()
    call void @use(ptr %p)
    ret ptr %p
}
```

If we inline `@foo` and propagate `nonnull` to `@bar`, it could change
the behavior of `@use` as instead of taking `null`, `@use` will
now be passed `poison`.

This can be even worth in a case like:
```
define nonnull ptr @foo() {
    %p = call noundef ptr @bar()
    ret ptr %p
}
```

Where propagating `nonnull` to `@bar` will cause UB on `null` return
of `@bar` (`noundef` + `poison`) where it previously wouldn't
have occurred.

To fix this, we only propagate poison generating return attributes if
either 1) The only use of the callsite to propagate too is return and
the callsite to propagate too doesn't have `noundef`. Or 2) the
callsite to be be inlined has `noundef`.

The former case ensures no new UB or `poison` values will be
added. The latter is UB anyways if the value is `poison` so we can go
ahead without worrying about behavior changes.

goldsteinn commented:
``

> It occurs to me that the current return attribute propagation is currently buggy for poison-generating attributes: https://llvm.godbolt.org/z/x8n18q9Mj
> 
> In this case the argument to use() will now be poison as well, while before inlining only the return value was poison.
> 
> This code needs to distinguish poison and UB generating attributes.

Fixed with new commit (`[Inliner] Fix bug...`).

jdoerfert commented:
I don't think 10 commit PRs are helpful. This should be 3 PRs at least. One is trivial, one is probably easy, and then one with the new stuff.

jdoerfert commented:
> > I don't think 10 commit PRs are helpful. This should be 3 PRs at least. One is trivial, one is probably easy, and then one with the new stuff.
> 
> So PR based on PR? The thing is there are dependencies between commits. My expectation is that I will push commits as they are approved, not the whole PR at once.

We always had dependences between commits. Just mention it in the commit message. 
This PR contains arguably different things. It simply makes it harder to review it. The guidance on this is clear, make review chunks as small if possible. The RFC commits, the fix, and the new functionality do not need to be commingled.

goldsteinn commented:
> > > I don't think 10 commit PRs are helpful. This should be 3 PRs at least. One is trivial, one is probably easy, and then one with the new stuff.
> > 
> > 
> > So PR based on PR? The thing is there are dependencies between commits. My expectation is that I will push commits as they are approved, not the whole PR at once.
> 
> We always had dependences between commits. Just mention it in the commit message. This PR contains arguably different things. It simply makes it harder to review it. The guidance on this is clear, make review chunks as small if possible. The RFC commits, the fix, and the new functionality do not need to be commingled.

Okay, I can't really see a reasonable way to create a proper series. I am going to submit this piece-meal.
At the moment only up to the bugfix submitted.
Do you by chance know a reasonable way to create a series of PRs s.t the individual commits can be
reasonably reviewed while also keeping all the commits together?
I look at SO answers like: https://stackoverflow.com/questions/30768148/git-split-pull-request-into-smaller-prs-based-upon-the-new-directories-in-the
but they only seem to work if the changes to to distinct files.

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1355,14 +1355,23 @@ static AttrBuilder IdentifyValidAttributes(CallBase &CB) {
+  AttrBuilder Valid(CB.getContext());
+  // Only allow these white listed attributes to be propagated back to the
+  // callee. This is because other attributes may only be valid on the call
+  // itself, i.e. attributes such as signext and zeroext.
```
Don't think we need to copy this comment.

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1355,14 +1355,23 @@ static AttrBuilder IdentifyValidAttributes(CallBase &CB) {
+  return Valid;
+}
+
+static AttrBuilder IdentifyValidPoisonGeneratingAttributes(CallBase &CB) {
```
We should rename the above function to ValidUBGeneratingAttributes for symmetry.

nikic replied:
Sorry, this was unclear because GitHub doesn't allow commenting on unchanged lines: I meant that IdentifyValidAttributes should become IdentifyValidUBGeneratingAttributes and this one should stay as IdentifyValidPoisonGeneratingAttributes.

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1407,6 +1416,58 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
     AttributeList AL = NewRetVal->getAttributes();
     AttributeList NewAL = AL.addRetAttributes(Context, Valid);
+    // Attributes that may generate poison returns are a bit tricky. If we
+    // propagate them, other uses of the callsite might have there behavior
```
there -> their

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1407,6 +1416,58 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
     AttributeList NewAL = AL.addRetAttributes(Context, Valid);
+    // Attributes that may generate poison returns are a bit tricky. If we
+    // propagate them, other uses of the callsite might have there behavior
+    // change/or cause UB (if they have noundef) b.c of the new potential
```
Drop `/`?

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1407,6 +1416,58 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
+        // necessarily have UB so don't propagate. TODO: This is overly
+        // conservative, we could check the uses to see if poison would change
+        // behavior / trigger UB.
+        for (const llvm::User *U : RetVal->users()) {
```
Does this do anything useful over a hasOneUse check? We only allow the case where the call and return are in the same block, so there can only be one ReturnInst user.

goldsteinn replied:
Not at the moment, I guess my feeling is we would like to be able to handle uses that can't cause UB (something like an `ptrmask` use) which will require looping. Secondly, this seems more "correct" should our constraints above change, and will be roughly the same cost as `hasOneUse` in the meantime (will be at most 2 iter).

goldsteinn replied:
See `caller12_todo` test as an example of what we should be able to do.

nikic replied:
IMHO we should only introduce this loop once it actually does something useful. For now this is just confusing dead code.

nikic reviewed (commented):
Second commit is missing test diffs -- either tests are missing or improperly committed.

goldsteinn commented:
> Second commit is missing test diffs -- either tests are missing or improperly committed.

Sorry, had gotten messed up when splitting. Fixed.

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1406,7 +1417,55 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
+    // In case 3, we can never propagate nonnull because it may create UB due to
+    // the noundef on @bar.
+    if (ValidPG.hasAttributes()) {
+      bool Okay = true;
+      if (!CB.hasRetAttr(Attribute::NoUndef)) {
+        // Cover case 3 where introducing a poison generating attribute can
+        // cause immediate UB.
+        Okay = !RetVal->hasRetAttr(Attribute::NoUndef);
+        // This is an overly conservative way to check that there are no other
+        // uses whose behavior may change/may introduce UB if we potentially
+        // return poison.
+        // TODO: Iterative through uses and only bail if we have a potentially
+        // dangerous use.
+        Okay &= RetVal->hasOneUse();
+      }
+      if (Okay)
```
Nit: I would write this either as
```suggestion
      // Comments...
      if (CB.hasRetAttr(Attribute::NoUndef) ||
          (!RetVal->hasRetAttr(Attribute::NoUndef) && RetVal->hasOneUse()))
```
or move it into a function with early returns, rather than using an `Okay` flag.

goldsteinn replied:
Done, although imo it makes the comments a bit more difficult to read.

nikic commented on llvm/lib/Transforms/Utils/InlineFunction.cpp:
```diff
@@ -1406,7 +1417,55 @@ static void AddReturnAttributes(CallBase &CB, ValueToValueMapTy &VMap) {
+        // This is an overly conservative way to check that there are no other
+        // uses whose behavior may change/may introduce UB if we potentially
+        // return poison.
+        // TODO: Iterative through uses and only bail if we have a potentially
```
Iterate

nikic commented on llvm/test/Transforms/Inline/access-attributes-prop.ll:
```diff
@@ -0,0 +1,498 @@
+declare void @bar1(ptr %p)
+declare void @bar2(ptr %p, ptr %p2)
+
+define dso_local void @foo1_rdonly(ptr readonly %p) {
```
`dso_local` shouldn't be needed.

nikic reviewed (approved):
LGTM

goldsteinn commented:
Pushed.
